rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1614 @@
+  # Good
+  {
+    foo = {
+      bar =
+        if baz == null then
+          10
+        else
+          20;
+    };
+  }
+  ```
+
+### Editor Config
+
+This [editor config](https://editorconfig.org/) specifies the basic details about Nix files:
+
+```editorconfig
+end_of_line = lf
+insert_final_newline = true
+trim_trailing_whitespace = true
+charset = utf-8
+indent_style = space
+```
+
+### Single-line common ancestor expression rule
+
+For any two (sub-)expressions that are fully on a common single line, their smallest common ancestor expression must also be on the same line.
+
+**Example**
+
+```nix
+# Bad, expressions cond and foo are fully on the same line,
+# but their smallest common ancestor expression is the entire if-then-else, which spans multiple lines
+if cond then foo
+else bar
+
+# Okay, cond, foo and bar have the if-then-else as a common ancestor expression,
+# which is also fully on the same line
+if cond then foo else bar
+
+# Bad, due to function application precedence, the smallest common ancestor expression
+# of foo and bar is `foo || bar baz`, which spans two lines
+foo || bar
+  baz
+```
+
+**Rationale**
+
+This rule has turned out to be very practical at catching code that could be potentially hard to understand or edit.
+
+### Line length
+
+- There should be a configurable _soft_ line length limit, limiting the number of characters on one line without counting the leading indentation.
+  The default should be 100 characters.
+- There may also be a configurable _hard_ line length limit, which includes the leading indentation.
+- String-like values such as strings, paths, comments, urls, etc. may go over the hard line length limit.
+
+### Indentation
+
+- Two spaces must be used for each indentation level.
+  - This may be revisited should Nix get proper support for [using tabs for indentation](https://github.com/NixOS/nix/issues/7834) in the future.
+- Vertical alignment must be ignored, both at the start of the line and within lines.
+  - Examples:
+    ```nix
+    {
+      # Bad, vertical alignment Within lines
+      linux    = { execFormat = elf;     families = {              }; };
+      netbsd   = { execFormat = elf;     families = { inherit bsd; }; };
+      none     = { execFormat = unknown; families = {              }; };
+      openbsd  = { execFormat = elf;     families = { inherit bsd; }; };
+
+      # Bad, vertical alignment at the start of line
+      optExecFormat =
+        lib.optionalString (kernel.name == "netbsd" &&
+                            gnuNetBSDDefaultExecFormat cpu != kernel.execFormat
+                           )
+                           kernel.execFormat.name;
+    }
+    ```
+- Increasing indentation levels must not be "skipped": On subsequent lines, indentation can only increase by at most one level, but may decrease arbitrarily many levels.
+  - Examples:
+    ```nix
+    buildInputs = [
+        foo # <-- Bad, indentation increases by 2 levels
+      ] // lib.optionals cond [
```

Comment:
Lists should be concatinated with `++`. Also I think the expression would require parentheses otherwise it would try to concatinate the function instead:
```
] ++ (lib.optionals cond [
```

Reply:
> Lists should be concatinated with `++`. Also I think the expression would require parentheses otherwise it would try to concatinate the function instead:
> 
> ```
> ] ++ (lib.optionals cond [
> ```

Right regarding `++`! Though the extra parenthesis are not required as is widely seen throughout Nixpkgs.